ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -106,105 +101,48 @@ public HelixClusterManager(ClusterMapConfig clusterMapConfig, String instanceNam
-            // provider should be already populated.
-            clusterChangeHandler.setRoutingTableSnapshot(routingTableProvider.getRoutingTableSnapshot());
-            // the initial routing table change should populate the instanceConfigs. If it's empty that means initial
-            // change didn't come and thread should wait on the init latch to ensure routing table snapshot is non-empty
-            if (clusterChangeHandler.getRoutingTableSnapshot().getInstanceConfigs().isEmpty()) {
-              // Periodic refresh in routing table provider is enabled by default. In worst case, routerUpdater should
-              // trigger routing table change within 5 minutes
-              logger.info("Routing table snapshot in {} is currently empty. Waiting for initial notification.", dcName);
-              clusterChangeHandler.waitForInitNotification();
-            }
-
-            if (!clusterMapConfig.clustermapListenCrossColo && manager != localManager) {
-              manager.disconnect();
-              logger.info("Stopped listening to cross colo ZK server {}", zkConnectStr);
-            }
-          } catch (Exception e) {
-            initializationFailureMap.putIfAbsent(dcName, e);
-          } finally {
-            initializationAttemptComplete.countDown();
-          }
-        }, false).start();
+        DatacenterInitializer initializer =
+            new DatacenterInitializer(clusterMapConfig, localManager, helixFactory, dcZkInfo, selfInstanceName,
+                partitionOverrideInfoMap, clusterChangeHandlerCallback, helixClusterManagerCallback,
+                helixClusterManagerMetrics, sealedStateChangeCounter, partitionMap, partitionNameToAmbryPartition,
+                ambryPartitionToAmbryReplicas);
+        initializer.start();
+        initializers.add(initializer);
       }
-      try {
-        initializationAttemptComplete.await();
-      } catch (InterruptedException e) {
-        initializationFailureMap.putIfAbsent(clusterMapConfig.clusterMapDatacenterName, e);
+      for (DatacenterInitializer initializer : initializers) {
+        try {
+          DcInfo dcInfo = initializer.join();
+          dcToDcInfo.put(dcInfo.dcName, dcInfo);
+          dcIdToDcName.put(dcInfo.dcZkInfo.getDcId(), dcInfo.dcName);
+        } catch (Exception e) {
+          initializationFailureMap.putIfAbsent(initializer.getDcName(), e);
+        }
+      }
+
+      for (DcInfo dcInfo : dcToDcInfo.values()) {
+        if (dcInfo.clusterChangeHandler instanceof ClusterMapChangeListener) {
+          ClusterMapChangeListener listener = (ClusterMapChangeListener) dcInfo.clusterChangeHandler;
+          registerClusterMapListener(listener);
+          List<ReplicaId> allReplicas = new ArrayList<>();
+          for (Set<AmbryReplica> replicas : ambryPartitionToAmbryReplicas.values()) {
+            allReplicas.addAll(replicas);
```

Comment:
That is true. I will do that! However, we should be careful in making it clear that this is specifically tailored to `CloudServiceClusterChangeHandler, since in most cases one would expect all replicas to be provided.